experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+
+            # if args are DTensors, get a local shard. But we also need to check
+            # that their placements match the intended placement.
+            for i in range(len(used_args)):
+                arg = used_args[i]
+                if isinstance(arg, DTensor):
+                    assert (
+                        arg.placements == schemas[i].placements and
+                        arg.device_mesh == schemas[i].mesh
+                    )
+                    used_args[i] = arg.to_local()
+
+            outputs = compiled_f(used_args)
+
+            return outputs if output_schemas is None else [
+                output if schema is None else DTensor.from_local(
+                    local_tensor=output,
+                    device_mesh=schema.mesh,
+                    placements=schema.placements,
+                    run_check=False,
+                )
+                for output, schema in zip(outputs, output_schemas)
+            ]
+
+
+    return inductor_fn
+
+
+# =====================================
+# 1. Minimal SGD example, no nn.Module
+# =====================================
+
+def inplace_test(x, p):
+    p.grad = None  # zero_grad
+    loss = (torch.sigmoid(torch.matmul(x, p)) ** 2).sum()
+    # pg, = torch.autograd.grad(loss, p)
+    loss.backward()
+    with torch.no_grad():
+        p -= 0.1 * p.grad
+    return (loss, )
+
+
+def main_simple():
+    fake_mode = FakeTensorMode()
+    torch.set_default_device('cuda')
+
+    if DIST:
+        mesh = DeviceMesh('cuda', range(torch.distributed.get_world_size()))
+        default_shard_schema = Schema(mesh, [Replicate()])
+
+    # with fake_mode:
+    with fake_mode:
+        x_fake = torch.randn(10, 2, device='cuda')
+        p_fake = torch.randn(2, 1, requires_grad=True, device='cuda')
+
+        if DIST:
+            mark_placements(x_fake, [Shard(0)])
+            mark_placements(p_fake, [Replicate()])
+
+            # x_shard = get_example_shard(x, mesh)
+            # dprint(f'Input shape: {x.shape}; shard shape: {x_shard.shape}')
+        else:
+            # x_shard = x
+            default_shard_schema = None
+
+        compiled_f = make_inductor_fn(
+            inplace_test,
+            [x_fake, p_fake],
+            default_shard_schema=default_shard_schema,
+        )
+
+    x = torch.randn(10, 2, device='cuda')
+    p = torch.randn(2, 1, requires_grad=True, device='cuda')
+
+    if DIST:
+        x = distribute_tensor(x, mesh, [Shard(0)])
+        p = distribute_tensor(p, mesh, [Replicate()])
+
+    if DIST:
+        # make sure our parameter is the same across devices initially
+        # doing this after compile because during compilation we're running the training step a few times
+        # TODO: fix this. we should run compilation with fake tensors.
+        # torch.distributed.all_reduce(p)
+        pass
+
+    dprint(f'after compile, param norm = {(p*p).sum()}')
+
+    with torch.no_grad():
```

Comment:
Do we always need to do this in the no_grad() context?

Reply:
Yes otherwise it tries to compute gradients of an inplace operation, namely the copy_ operation that happens at parameter update time, and it will complain that you can't do backwards for an inplace operation. This is fine because at this point, backwards is already baked into the graph.

In case you want to do backwards of backwards i think you could use functorch beforehand.